bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/DataLayerFactory.java

```diff
@@ -0,0 +1,144 @@
+import java.util.*;
+
+public class DataLayerFactory {
+
+    static public Map<String, Object> build(DataLayerProvider provider) {
+        MapWrapper data = new MapWrapper();
+        data.put("id", invoke(provider, "getDataLayerId"));
+        data.put("type", invoke(provider, "getDataLayerType"));
+        data.put("name", invoke(provider, "getDataLayerName"));
+        data.put("path", invoke(provider, "getDataLayerPath"));
+        data.put("title", invoke(provider, "getDataLayerTitle"));
+        data.put("template", invoke(provider, "getDataLayerTemplate"));
+        // data.put("src", invoke(provider, "getDataLayerSrc"));
+        data.put("text", invoke(provider, "getDataLayerText"));
+        data.put("tags", invoke(provider, "getDataLayerTags"));
+        data.put("asset", getAssetMetadata(provider));
+        data.put("linkUrl", invoke(provider, "getDataLayerLinkUrl"));
+        data.put("language", invoke(provider, "getDataLayerLanguage"));
+        data.put("itemsCount", invoke(provider, "getDataLayerItemsCount"));
+        data.put("activeItem", invoke(provider, "getDataLayerActiveItem"));
+        data.put("expandedItems", getExpandedItems(provider));
+        return data.getMap();
+    }
+
+    static private Object invoke(DataLayerProvider provider, String method) {
+        try {
+            Method providerMethod = provider.getClass().getMethod(method);
+            Object result = providerMethod.invoke(provider);
+            return result;
+        } catch (UnsupportedOperationException | NoSuchMethodException | IllegalAccessException
+                | InvocationTargetException e) {
+            return null;
+        }
+    }
+
+    static private Map<String, Object> getAssetMetadata(DataLayerProvider provider) {
+        Map<String, Object> assetMetadataObject = null;
+        Resource assetResource = null;
+        try {
+            assetResource = provider.getAssetResource();
+        } catch (UnsupportedOperationException e) {
+            // do nothing as we need to continue
+        }
+        if (assetResource != null) {
+            Asset asset = assetResource.adaptTo(Asset.class);
+            if (asset != null) {
+                MapWrapper assetMetadata = new MapWrapper();
+                assetMetadata.put("id", asset.getID());
+                assetMetadata.put("name", asset.getName());
+                assetMetadata.put("path", asset.getPath());
+                assetMetadata.put("type", asset.getMimeType());
+                assetMetadata.put("tags", getAssetTags(asset));
+                assetMetadata.put("modifyDate", getAssetLastModifiedDate(asset, assetResource));
+                assetMetadataObject = assetMetadata.getMap();
+            }
+        }
+        return assetMetadataObject;
+    }
+
+    static private Map<String, Object> getAssetTags(Asset asset) {
+        MapWrapper assetTags = new MapWrapper();
+        String tagsValue = asset.getMetadataValueFromJcr("cq:tags");
```

Comment:
check for null?

Reply:
`getAssetTags` is called from `getAssetMetadata` where the asset is checked for `null`